providers/implementations/keymgmt/ecx_kmgmt.c

```diff
@@ -224,10 +390,421 @@ static const OSSL_PARAM *ecx_gettable_params(void)
         { OSSL_FUNC_KEYMGMT_IMPORT_TYPES, (void (*)(void))ecx_imexport_types }, \
         { OSSL_FUNC_KEYMGMT_EXPORT, (void (*)(void))ecx_export }, \
         { OSSL_FUNC_KEYMGMT_EXPORT_TYPES, (void (*)(void))ecx_imexport_types }, \
+        { OSSL_FUNC_KEYMGMT_GEN_INIT, (void (*)(void))alg##_gen_init }, \
+        { OSSL_FUNC_KEYMGMT_GEN, (void (*)(void))alg##_gen }, \
+        { OSSL_FUNC_KEYMGMT_GEN_CLEANUP, (void (*)(void))ecx_gen_cleanup }, \
         { 0, NULL } \
     };
 
 MAKE_KEYMGMT_FUNCTIONS(x25519)
 MAKE_KEYMGMT_FUNCTIONS(x448)
 MAKE_KEYMGMT_FUNCTIONS(ed25519)
 MAKE_KEYMGMT_FUNCTIONS(ed448)
+
+#ifdef S390X_EC_ASM
+# include "s390x_arch.h"
+# include "internal/constant_time.h"
+
+static void s390x_x25519_mod_p(unsigned char u[32])
```

Comment:
All this seems to be duplicated from ecx_meth.c. Do we really have to have 2 identical copies of this code?

Reply:
I'd ideally like remove the code in `ecx_meth.c`.  I'm not sure how that plays with legacy support -- the functionality won't change but performance will.  There is additional code in `ecx_meth.c` which we don't want to carry across, so things needed to be split.

The arguments to the top level functions were different, so some code needs duplication.

I can look into splitting off some the code in libcrypto instead rather than slating it for removal from `ecx_meth.c`.

Reply:
I think the code that needs to be in both places should be split out into a separate file. The relevant functions can be defined in some internal header file, and the code should be included in both libcrypto and the fips provider.

Reply:
Working on it...

Reply:
It's in now.